### PR TITLE
Added PipelineParser that can integrate several parsers together

### PR DESCRIPTION
**Pull request by smilesrg**
Added PipelineParser that can integrate several parsers together and execute them one by one.

### CODE REVIEW

**Comment by tarjei** [comment]
Looks good!

File: Markup/PipelineParser.php [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+/**
+ * Pipeline class for parsers.
+ *
+ * @author Sergii Smirnov <free.smilesrg@gmail.com>
+ */
+class PipelineParser implements ParserInterface
+{
+    /** @var array ParserInterface */
+    protected $pipeline = array();
```
Comment (stof) [inline Markup/PipelineParser.php]: please make this private

File: Markup/PipelineParser.php [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+namespace FOS\CommentBundle\Markup;
+
+/**
+ * Pipeline class for parsers.
+ *
+ * @author Sergii Smirnov <free.smilesrg@gmail.com>
+ */
+class PipelineParser implements ParserInterface
+{
+    /** @var array ParserInterface */
```
Comment (stof) [inline Markup/PipelineParser.php]: should be ``@var ParserInterface[]``

File: Resources/doc/9e-integrating_several_parsers.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        class: MyBundle\Parser\MyParser
+```
+
+Then, you have to define PipelineParser as a service and configure it with parsers.
+
+```yaml
+    markuper.pipeline_parser:
+        class: FOS\CommentBundle\Markup\PipelineParser
+        calls:
+            - [addToPipeline, ["@markup.exercise_html_purifier"]]
```
Comment (stof) [inline Resources/doc/9e-integrating_several_parsers.md]: the HTML purifier should probably be applied *after* your own parser though most of the time, so I would show the other order

File: Resources/doc/9e-integrating_several_parsers.md [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+```yaml
+    # See https://github.com/FriendsOfSymfony/FOSCommentBundle/blob/master/Resources/doc/9a-markup_htmlpurifier.md
+    markup.exercise_html_purifier:
+        class: FOS\CommentBundle\Markup\HtmlPurifier
+        arguments: [ @exercise_html_purifier.default ]
+        
+    # Your own parser    
+    markup.my_parser:
+        #This is your own parser that implements ParserInterface
+        class: MyBundle\Parser\MyParser
```
Comment (stof) [inline Resources/doc/9e-integrating_several_parsers.md]: I would mark these services as private

File: Resources/doc/9e-integrating_several_parsers.md [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        class: FOS\CommentBundle\Markup\PipelineParser
+        calls:
+            - [addToPipeline, ["@markup.exercise_html_purifier"]]
+            - [addToPipeline, ["@markup.my_parser"]]
+```
+
+The order of parsers is important. First parser that has been added to pipeline is executed first.
+
+
+The last step is to use pipeline parser as described in [documentation](https://github.com/FriendsOfSymfony/FOSCommentBundle/blob/master/Resources/doc/9-using_a_markup_parser.md)
```
Comment (stof) [inline Resources/doc/9e-integrating_several_parsers.md]: please use a relative link instead, to ensure that the right version of the doc is used when looking at the doc in other versions than the master branch. github handles relative links properly

File: Resources/doc/9e-integrating_several_parsers.md [hunk 6]
```diff
@@ -0,0 +1,45 @@
+Step 9e: Integrating several parsers together
+=============================================
+
+Purpose
+--------
+Imagine that you need to integrate ExerciseHTMLPurifierBundle parser, bbcode parser, any other parser... it's better to have several parsers where each is responsible for it's action. That's why it's better to make chain of parsers, where order is important and parsers works one after other.
```
Comment (stof) [inline Resources/doc/9e-integrating_several_parsers.md]: typo: ``its action``

File: Resources/doc/9e-integrating_several_parsers.md [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+
+Usage
+-----
+
+First, configure parser classes as services. You don't have to use all parsers that this bundle provides. The main parser of this bundle is PipelineParser which provides pipeline for parser classes.
```
Comment (stof) [inline Resources/doc/9e-integrating_several_parsers.md]: ``this bundle``looks weird here. It was fine in the doc of your own bundle where it was the main feature, but it is not the main feature anymore when adding it in FOSCommentBundle itself

File: Resources/doc/9e-integrating_several_parsers.md (hunk 4 above)
Comment (smilesrg) [inline Resources/doc/9e-integrating_several_parsers.md]: Yep! That was in PipelineParserBundle, I've just lost it here. Thanks!

File: Resources/doc/9e-integrating_several_parsers.md (hunk 3 above)
Comment (smilesrg) [inline Resources/doc/9e-integrating_several_parsers.md]: I made UserMentionParser and HashtagParser that implements HTML code in original comment, that's why HTML code should be stripped first.

**Comment by smilesrg** [comment]
@stof I've fixed code review comments. Thank you!

File: Resources/doc/9e-integrating_several_parsers.md (hunk 3 above)
Comment (stof) [inline Resources/doc/9e-integrating_several_parsers.md]: well, it depends what kind of replacement they do. Purifying at the end of the pipeline is safer as it avoids XSS issues (if you purify before the extra processing, it opens the door to XSS issues thanks to edge cases of the processing)

File: Resources/doc/9e-integrating_several_parsers.md [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+Usage
+-----
+
+First, configure parser classes as services.
+
+```yaml
+    # See 9a-markup_htmlpurifier.md
+    markup.exercise_html_purifier:
+        class: FOS\CommentBundle\Markup\HtmlPurifier
+        arguments: [ @exercise_html_purifier.default ]
```
Comment (stof) [inline Resources/doc/9e-integrating_several_parsers.md]: this one should also be private (even ``markuper.pipeline_parser`` can be private, as FOSCommentBundle will never retrieve it dynamically using this name)

File: Resources/doc/9e-integrating_several_parsers.md [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+```yaml
+    markuper.pipeline_parser:
+        class: FOS\CommentBundle\Markup\PipelineParser
+        calls:
+            - [addToPipeline, ["@markup.exercise_html_purifier"]]
+            - [addToPipeline, ["@markup.my_parser"]]
+```
+
+The order of parsers is important. First parser that has been added to pipeline is executed first.
```
Comment (stof) [inline Resources/doc/9e-integrating_several_parsers.md]: ``to the pipeline``.

And I think we could rephrase the whole sentence:

```
Parsers are executed in the order in which they are added to the pipeline.
```

What do you think about it ?

Comment (smilesrg) [inline Resources/doc/9e-integrating_several_parsers.md]: Good idea! The sentence has been replaced.

File: Resources/doc/9e-integrating_several_parsers.md (hunk 8 above)
Comment (smilesrg) [inline Resources/doc/9e-integrating_several_parsers.md]: Done.

**Comment by smilesrg** [comment]
@stof - would you please accept this PR?

**Comment by XWB** [comment]
@smilesrg There is still a type in the docs.

File: Resources/doc/9e-integrating_several_parsers.md (hunk 6 above)
Comment (XWB) [inline Resources/doc/9e-integrating_several_parsers.md]: Ping @smilesrg

Comment (skylord123) [inline Resources/doc/9e-integrating_several_parsers.md]: Also, `works one after other` should be `works one after another`

Comment (smilesrg) [inline Resources/doc/9e-integrating_several_parsers.md]: Done

**Comment by smilesrg** [comment]
@stof Fixed typos. Feel free to re-write docs, I've written is as best as I could.

**Comment by XWB** [comment]
@smilesrg Something went wrong, there are too many commits. Did you rebase from master?

**Comment by smilesrg** [comment]
Yep, I did something wrong. Just tried to update to latest master. As far
as I remember, tried to rebase somehow.

10 Май 2017 г. 10:04 пользователь "Karel" <notifications@github.com>
написал:

**Comment by XWB** [comment]
@smilesrg You need to rebase from master branch.

**Comment by smilesrg** [comment]
@XWB Don't really remember what I did with this merging and rebasing, but I tried again. If it's still wrong, then I propabaly will create a new form and apply commiths there.

**Comment by smilesrg** [comment]
@stof @XWB closing this PR, created new one #606

**Pull request closed without merging**